// @appland/appmap-agent-js/dist/bundles/error.mjs
import { readFile } from 'node:fs/promises';
import { URL as URL$1 } from 'node:url';

const toAbsoluteUrl = (relative, base_url) =>
  new URL$1(
    /^[a-zA-Z]:\/[^/]/u.test(relative) ? `/${relative}` : relative,
    base_url,
  ).href;

// Consistent way to retreive home url in prod and test.

const {
  URL,
  JSON: { parse: parseJSON },
} = globalThis;

let url = toAbsoluteUrl(".", import.meta.url);

while (!url.endsWith("appmap-agent-js/")) {
  url = toAbsoluteUrl("..", url);
}

const self_directory = url;

const self_package = parseJSON(
  await readFile(
    new URL(toAbsoluteUrl("package.json", self_directory)),
    "utf8",
  ),
);

const version = self_package.version;

const { Error } = globalThis;

const issues = "https://github.com/getappmap/appmap-agent-js/issues";
const slack = "https://appmap.io/slack";
const documentation =
  "https://github.com/getappmap/appmap-agent-js/blob/main/REFERENCE.md";
const instruction = "https://appmap.io/docs/appmap-overview.html";

const internal_message = `
[appmap@${version}] Detected an internal appmap error.\
 This is probably an issue with how the AppMap agent is being used.\
 Please consider submitting a bug report at:
  ${issues}

`;

const external_message = `
[appmap@${version}] Detected an external appmap error.\
 This is probably an issue with how the AppMap agent is being used.
- You can look for answers in our installation instructions:
    ${instruction}
- We also have a reference documentation:
    ${documentation}
- You can ask for help in our public slack channel:
    ${slack}

`;

const unknown_message = `
[appmap@${version}] Detected an unknown error.\
 If this error disapear when not recording your application, it is probably an issue within the appmap framework.\
 If this is the case, please consider submitting a bug report at:
  ${issues}

`;

class AppmapError extends Error {}

class InternalAppmapError extends AppmapError {
  constructor(message) {
    super(message);
    this.name = "InternalAppmapError";
  }
}

class ExternalAppmapError extends AppmapError {
  constructor(message) {
    super(message);
    this.name = "ExternalAppmapError";
  }
}

const reportError = (error) => {
  if (error.name === "InternalAppmapError") {
    return internal_message;
  } else if (error.name === "ExternalAppmapError") {
    return external_message;
  } else {
    return unknown_message;
  }
};

export { AppmapError, ExternalAppmapError, InternalAppmapError, reportError };
